sql/sql_plugin.cc

```diff
@@ -4028,45 +4047,69 @@ int unlock_plugin_data()
   Create deep copy of system_variables instance.
 */
 struct system_variables *
-copy_system_variables(const struct system_variables *src,
-                      bool enable_plugins)
+copy_system_variables(THD *thd, bool enable_plugins)
 {
   struct system_variables *dst;
+  ulong idx;
 
-  DBUG_ASSERT(src);
+  DBUG_ASSERT(thd);
 
   dst= (struct system_variables *)
     my_malloc(sizeof(struct system_variables), MYF(MY_WME | MY_FAE));
-  *dst = *src;
+  *dst = thd->variables;
 
   if (dst->dynamic_variables_ptr)
```

Comment:
I am not sure if this block should be protected by `LOCK_system_variables_hash`, in `intern_sys_var_ptr` I see
```
    thd->variables.dynamic_variables_ptr= (char*)
      my_realloc(thd->variables.dynamic_variables_ptr,
                 global_variables_dynamic_size,
                 MYF(MY_WME | MY_FAE | MY_ALLOW_ZERO_PTR));
```
is done under `LOCK_system_variables_hash` and it looks logical to me. But current set-statement implementation seems doesn't bother to acquire this lock at all.